app/controllers/carto/api/public/federated_tables_controller.rb

```
@@ -0,0 +1,285 @@
+            ensure_no_extra_request_params(ALLOWED_PUT_FEDERATED_SERVER_ATTRIBUTES)
+          end
+        end
+
+        def load_remote_table_attributes
+          @remote_table_attributes = params.slice(*REMOTE_TABLE_ATTRIBUTES)
+          @remote_table_attributes[:local_table_name_override] ||= @remote_table_attributes[:remote_table_name]
+        end
+
+        def load_remote_table
+          @remote_table = @service.get_remote_table(
+            federated_server_name: params[:federated_server_name],
+            remote_schema_name: params[:remote_schema_name],
+            remote_table_name: params[:remote_table_name]
+          )
+          raise Carto::LoadError.new("Table '#{params[:remote_schema_name]}'.'#{params[:remote_table_name]}' not found at '#{params[:federated_server_name]}'") if @remote_table.nil?
+        rescue Sequel::DatabaseError => exception
+          handle_service_error(exception)
+        end
+
+        def check_remote_table
+          raise Carto::LoadError.new("Remote table key not found: #{params[:federated_server_name]}/#{params[:remote_schema_name]}.#{params[:remote_table_name]}") unless @remote_table
+        end
+
+        def ensure_required_remote_table_attributes
+          if request.post?
+            ensure_required_request_params(REQUIRED_POST_REMOTE_TABLE_ATTRIBUTES)
+          else
+            ensure_required_request_params(REQUIRED_PUT_REMOTE_TABLE_ATTRIBUTES)
+            ensure_no_extra_request_params(ALLOWED_PUT_REMOTE_TABLE_ATTRIBUTES, 422)
+          end
+        end
+
+        def ensure_readonly_mode
+            raise Carto::UnprocesableEntityError.new("Invalid access mode: '#{params[:mode]}'. Only 'read-only' accepted") unless ("read-only".casecmp params[:mode]) == 0
+        end
+
+        def check_permissions
+          @api_key = Carto::ApiKey.find_by_token(params["api_key"])
+          raise UnauthorizedError unless @api_key.master? || @api_key.dataset_metadata_permissions
```

What happens if it does not find the api_key or I send an api_key of other user?

---

Ultimately permissions are enforced by DB and roles. The master api_key user's role is used to create a role to manage all FT's dependent objects. This API checks for DB errors, so I think it is pretty well covered overall.

This approach avoids duplication of ACL's in rails and in DB.

Please correct me if I'm wrong

---

Thanks @rafatower that's a good approach.

I was asking about the validation in this specific method, even when the permissions are enforce in the DB roles, if I send an invalid api_key we could raise an error directly from the API. A nice to have anyway :)